gateways/gateway.go

```
@@ -79,10 +115,211 @@ func (gc *GatewayConfig) newConfigMapWatch(name string) *cache.ListWatch {
+	gc.log.Info().Interface("new-config-keys", newConfigKeys).Msg("new config")
+
+	// run new configurations
+	for _, newConfigKey := range newConfigKeys {
+		newConfig := newConfigs[newConfigKey]
+		gc.registeredConfigs[newConfigKey] = newConfig
+		go gc.runConfig(newConfigKey, newConfig)
+	}
+
+	gc.log.Info().Interface("registered-configs", gc.registeredConfigs).Msg("registered configs")
+
+	// remove stale configurations
+	for _, staleConfigKey := range staleConfigKeys {
+		staleConfig := gc.registeredConfigs[staleConfigKey]
+		gc.log.Info().Str("config", staleConfig.src).Msg("deleting config")
+		staleConfig.conn.Close()
+		delete(gc.registeredConfigs, staleConfigKey)
+	}
+	return nil
+}
+
+// runConfig establishes connection with gateway server and sends new configurations to run.
+// also disconnect the clients for stale configurations
+func (gc *gatewayConfig) runConfig(hash uint64, config *configData) error {
+
+	var opts []grpc.DialOption
+	opts = append(opts, grpc.WithInsecure())
+	conn, err := grpc.Dial(fmt.Sprintf(serverAddr, gc.rpcPort), opts...)
+	if err != nil {
+		gc.log.Fatal().Str("config-key", config.src).Err(err).Msg("failed to dial")
+	}
+	defer conn.Close()
+
+	// create a client for gateway server
+	client := gwProto.NewGatewayExecutorClient(conn)
+
+	eventStream, err := client.RunGateway(context.Background(), &gwProto.GatewayConfig{
+		Config: config.config,
+		Src:    config.src,
+	})
+
+	gc.log.Info().Str("config-key", config.src).Msg("connected with server and got a event stream")
+
+	if err != nil {
+		gc.log.Error().Str("config-key", config.src).Err(err).Msg("failed to get event stream")
+		return err
+	}
+
+	config.conn = conn
+
+	for {
+		event, err := eventStream.Recv()
+		if err == io.EOF {
+			gc.log.Info().Str("config-key", config.src).Msg("event stream stopped")
+			return nil
+		}
+		if err != nil {
+			gc.log.Warn().Str("config-key", config.src).Err(err).Msg("failed to receive events on stream")
+			return nil
```

should return the err